### ISSUE TITLE
Daemon closes the connection after request

### ISSUE THREAD

**Issue by ramal-ratne** [docker/docker#18114]
Description of problem:

I have searched existing issues on SIGPIPE and SO_KEEPALIVE. Although nothing pops up. Hence reporting as a new issue.

I am using **Docker Remote Api** to start a container and then inspect the container.  However the daemon has already closed the connection (SIGPIPE raised) from its end after the first request.

I have tried the **SO_KEEPALIVE** to keep the link open. Although that does not seem to have an effect.

The Docker Info
```
Containers: 1
Images: 92
Server Version: 1.9.0-rc4
Storage Driver: overlay
 Backing Filesystem: extfs
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 4.3.0-040300-generic
Operating System: Ubuntu 14.04.3 LTS
CPUs: 2
Total Memory: 1.756 GiB
```

uname -a:
```
Linux pyrus 4.3.0-040300-generic #201511020949 SMP Mon Nov 2 14:50:44 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux
```

Environment
```
Google Compute Engine machine.
```

Steps to re-produce 
```
1. Send start command via the API
2. Send a second command
3. Observe the SIGPIPE indicating the connection has been closed by the Docker server.
```

The strace.

**Open the socket**
```
socket(PF_LOCAL, SOCK_STREAM, 0)        = 3
connect(3, {sa_family=AF_LOCAL, sun_path="/var/run/docker.sock"}, 22) = 0
fcntl(3, F_SETFL, O_RDONLY|O_NONBLOCK)  = 0
setsockopt(3, SOL_SOCKET, SO_REUSEADDR, [1], 4) = 0
setsockopt(3, SOL_SOCKET, SO_KEEPALIVE, [1], 4) = 0
```

**Request start container**
```
sendto(3, "POST /containers/20ba90de12b3/start HTTP/1.2\r\nUser-Agent: hello/1.0\r\nAccept: */*\r\n\r\n\r\n", 86, 0, NULL, 0) = 86
recvfrom(3, 0x10cd01c, 86, 0, 0, 0)     = -1 EAGAIN (Resource temporarily unavailable)
rt_sigprocmask(SIG_BLOCK, [CHLD], [], 8) = 0
rt_sigaction(SIGCHLD, NULL, {SIG_DFL, [], 0}, 8) = 0
rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = 0
nanosleep({1, 0}, 0x7fffca765e80)       = 0
```
**Response OK - started**
```
recvfrom(3, "HTTP/1.1 204 No Content\r\nServer: Docker/1.9.0-rc4 (linux)\r\nDate: Fri, 20 Nov 2015 06:09", 87, 0, NULL, NULL) = 87
recvfrom(3, "", 0, 0, NULL, NULL)       = 0
fstat(1, {st_dev=makedev(0, 14), st_ino=11, st_mode=S_IFCHR|0620, st_nlink=1, st_uid=8868, st_gid=5, st_blksize=1024, st_blocks=0, st_rdev=make
dev(136, 8), st_atime=2015/11/20-06:09:44, st_mtime=2015/11/20-06:09:44, st_ctime=2015/11/19-13:32:39}) = 0
mmap(NULL, 4096, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7fa352ec2000
write(1, "TTP/1.1 204 No Content\r\n", 24) = 24
write(1, "Server: Docker/1.9.0-rc4 (linux)\r\n", 34) = 34
write(1, "Date: Fri, 20 Nov 2015 06:09\n", 29) = 29
```
**Request Inspection - However connection closed by the Daemon**
```
sendto(3, "TTP/1.1 204 No Content\r\nServer: Docker/1.9.0-rc4 (linux)\r\nDate: Fri, 20 Nov 2015 06:09GET /containers/20ba90de12b3/json HTTP/1.2
\r\nUser-Agent: hello/1.0\r\nAccept: */*\r\n\r\n\r\n", 170, 0, NULL, 0) = -1 EPIPE (Broken pipe)
--- SIGPIPE {si_signo=SIGPIPE, si_code=SI_USER, si_pid=13661, si_uid=0} ---
write(1, "SIGPIPE!! the other end closed the connection\n", 46) = 46
exit_group(1)                           = ?
+++ exited with 1 +++
```

The code.
```
#include <sys/socket.h>
#include <sys/types.h>
#include <netinet/in.h>
#include <netdb.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <unistd.h>
#include <fcntl.h>
#include <errno.h>
#include <signal.h>
#include <arpa/inet.h> 
#include <sys/un.h>

#define MAX_FRAME_BUFFER_SIZE	256

#define CONSTRUCT_PACKET(str) this_session->component_length = strlen(str); \
	strcpy(this_session->buffer + this_session->full_length, str); \
	this_session->full_length += this_session->component_length

typedef struct session {
	int sockfd;
	int full_length;
	int component_length;
	char buffer[MAX_FRAME_BUFFER_SIZE];
} session;

void sigpipe_handler(int sig0) {
	printf("SIGPIPE!! the other end closed the connection\n");
	 exit(1); 
}

void end_with_error(char *errorMessage) { 
	perror(errorMessage); 
	exit(1); 
}

session* open_session() {
	session* this_session = malloc(sizeof(session));
	this_session->sockfd = socket(AF_UNIX, SOCK_STREAM, 0);

	if(this_session->sockfd  < 0)
	{
		end_with_error("\n Error : Could not create socket \n");
	} 

	struct sockaddr_un serv_addr;
	serv_addr.sun_family = AF_UNIX;
	strcpy(serv_addr.sun_path, "/var/run/docker.sock");

	int rc = connect(this_session->sockfd, (struct sockaddr *)&serv_addr, 
		strlen(serv_addr.sun_path) + sizeof(serv_addr.sun_family));

	if( rc < 0)
	{
		end_with_error("\n Error : Connect Failed \n"); 
	}

	if ( fcntl(this_session->sockfd, F_SETFL, O_NONBLOCK) < 0) {
		end_with_error("\n Error : fcntl Failed \n"); 
	}

	int optval = 1;
	socklen_t optlen = sizeof(optval);

	if (setsockopt(this_session->sockfd, SOL_SOCKET, SO_REUSEADDR, &optval, optlen) < 0) 
	{
		end_with_error("\n Error : setsockopt SO_REUSEADDR Failed \n"); 
	}
   
	if (setsockopt(this_session->sockfd, SOL_SOCKET, SO_KEEPALIVE, &optval, optlen) < 0) 
	{
		end_with_error("\n Error : setsockopt SO_KEEPALIVE Failed \n"); 
	}
	
	return this_session;
}

void close_session(session* this_session)
{
	close(this_session->sockfd);
	free(this_session);
}
 
void build_http_frame(session* this_session)
{
	CONSTRUCT_PACKET( " HTTP/1.2\r\n" );
	CONSTRUCT_PACKET( "User-Agent: hello/1.0\r\n" );
	CONSTRUCT_PACKET("Accept: */*\r\n" );
	CONSTRUCT_PACKET( "\r\n" );
	CONSTRUCT_PACKET( "\r\n" );

	/* add "\r\n" at the header end */
}

int send_frame(session* this_session) 
{
	char* pos = this_session->buffer;
	int n_written = 0, remaining = this_session->full_length;

	while (remaining) {
		n_written = send(this_session->sockfd, pos, remaining, 0);

		if (n_written <= 0) {
			if (errno != EAGAIN) {
				end_with_error("\n Error : send Failed \n");
			}
		}
		else {
			remaining -= n_written;
			pos += n_written;
		}
	}

	return pos - this_session->buffer;
}

int recv_frame(session* this_session) 
{
	char* pos = this_session->buffer;
	int n_read = 0, remaining = this_session->full_length;

	while (1) {
        n_read = recv(this_session->sockfd, pos, remaining, 0);
        if (n_read == 0) {
            break;
        } else if (n_read < 0) {
			if (errno != EAGAIN) {
				end_with_error("\n Error : recv Failed \n");
			}
			else {
				sleep(1);
			}
        }
		remaining -= n_read;
		pos += n_read;
    }

	return pos - this_session->buffer;
}

int print_frame(session* this_session) 
{
	puts(this_session->buffer);
}

int start_container(session* this_session, const char* id) 
{
	CONSTRUCT_PACKET("POST ");
	CONSTRUCT_PACKET("/containers/");
	CONSTRUCT_PACKET(id);
	CONSTRUCT_PACKET("/start");

	build_http_frame(this_session);

	send_frame(this_session);
	recv_frame(this_session);
	print_frame(this_session);
}

int inspect_container(session* this_session, const char* id) 
{
	CONSTRUCT_PACKET("GET ");
	CONSTRUCT_PACKET("/containers/");
	CONSTRUCT_PACKET(id);
	CONSTRUCT_PACKET("/json");

	build_http_frame(this_session);

	send_frame(this_session);
	recv_frame(this_session);
	print_frame(this_session);
}

int main(int argc, char *argv[])
{
	signal(SIGPIPE, sigpipe_handler);

	session* this_session = open_session();

	start_container(this_session, argv[1]);
	inspect_container(this_session, argv[1]);

	close_session(this_session);

    return 0;
}
```

To compile and run
```
gcc -g -O0 test.c -o test
./test <container-id> 
```

**Comment by rajitha-wijayaratne** [comment]
I have updated the information as requested.  Thank you.

**Comment by crosbymichael** [comment]
Are you making sure you are creating the container first before starting it?

**Comment by rajitha-wijayaratne** [comment]
I am positive the container is running.

**Comment by LK4D4** [comment]
Never saw this since. Feel free to open new issue if problem still persists.
